File: 47-permutations-ii/47-permutations-ii.py
class Solution(object):
    def permuteUnique(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        nums.sort()
        used = dict()
        n = len(nums)
        result, track = [], []
        def generate(step):
            if step == n :
                result.append(track[:])
                return
            for i in range(n):
                if i > 0 and nums[i-1] == nums[i] and i-1 not in used:
                    continue
                if i in used:
                    continue
                track.append(nums[i])
                used[i] = 1
                generate(step+1)
                del used[i]
                track.pop()
        generate(0)
        return result

File: 47-permutations-ii/47-permutations-ii.java
class Solution {
    private Set<Integer> visited = new HashSet<>();
    public List<List<Integer>> permuteUnique(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> res = new LinkedList<>();
        permuteRes(nums, res, new LinkedList<>());
        return res;
    }
    
    public void permuteRes(int[] nums, List res, LinkedList temp){
        if(temp.size() == nums.length){
            res.add((List) temp.clone());
            return;
        }
        for(int i = 0; i < nums.length; i++){
            if(visited.contains(i)) continue;
            if(i > 0 && nums[i] == nums[i-1] && !visited.contains(i-1))continue;
            visited.add(i);
            temp.addLast(nums[i]);
            permuteRes(nums, res, temp);
            visited.remove(i);
            temp.pollLast();
        }
        
        
        
        
    }
}